maro/simulator/core.py

```diff
@@ -286,31 +278,30 @@ def _simulate(self):
                 if actions is not None and not isinstance(actions, Iterable):
                     actions = [actions]
 
-                # Generate a new atom event first.
-                action_event = self._event_buffer.gen_action_event(self._tick, actions)
-
-                # NOTE: decision event always be a CascadeEvent
-                # We just append the action into sub event of first pending cascade event.
-                pending_events[0].state = EventState.EXECUTING
-                pending_events[0].add_immediate_event(action_event, is_head=True)
-
-                if self._decision_mode == DecisionMode.Joint:
-                    # For joint event, we will disable following cascade event.
-
-                    # We expect that first action contains a src_event_id to support joint event with sequential action.
-                    action_related_event_id = None if len(
-                        actions) == 1 else getattr(actions[0], "src_event_id", None)
-
-                    # If the first action has a decision event attached, it means sequential action is supported.
-                    is_support_seq_action = action_related_event_id is not None
-
-                    if is_support_seq_action:
-                        for i in range(1, pending_event_length):
-                            if pending_events[i].id == actions[0].src_event_id:
-                                pending_events[i].state = EventState.FINISHED
-                    else:
-                        for i in range(1, pending_event_length):
-                            pending_events[i].state = EventState.FINISHED
+                if self._decision_mode == DecisionMode.Sequential:
+                    # Generate a new atom event first.
+                    action_event = self._event_buffer.gen_action_event(self._tick, actions)
+
+                    # NOTE: decision event always be a CascadeEvent
+                    # We just append the action into sub event of first pending cascade event.
+                    pending_events[0].state = EventState.EXECUTING
+                    pending_events[0].add_immediate_event(action_event, is_head=True)
+                else:
+                    # For joint mode, we will assign actions from beginning to end.
+                    # Then mark others pending events to finished if not sequential action mode.
+                    for i, pending_event in enumerate(pending_events):
+                        if i >= len(actions):
+                            if self._decision_mode == DecisionMode.Joint:
+                                # Ignore following pending events that have no action matched.
+                                pending_event.state = EventState.FINISHED
+                        else:
+                            # Set the state as executing, so event buffer will not pop them again.
```

Comment:
pop them **again**?

Reply:
same as joint mode, that pop all pending decision event at one time, not sequentially, it expect an action list that size is same as decision event. With this change, joint mode can accept actions one by one, and assign the to decision events in the same order, until all decision events get an action.